### PR TITLE
[dune] Add quickide target for building of IDE.

### PR DESCRIPTION
**Pull request by ejgallego**
Note that `states` doesn't work reliably yet, but that is a separate
problem that will be fixed in Dune 1.6.

[Or we could also fix it improving the rules in envars.ml]

### CODE REVIEW

**Comment by gares** [comment]
Thanks!
Just a minor remark: why "quickide" and not "ide" or "coqide"? What makes it quicker?

**Review by gares — APPROVED** [review]

**Comment by ejgallego** [comment]
I named it in accordance to the other targets, the idea is that in the future we should be way faster and just build the minimal needed stuff. As coqide is built by default I thought to use a more marked naming. We can rename the rule of course..

**Comment by ejgallego** [comment]
I named it in accordance to the other targets, the idea is that in the future we should be way faster and just build the minimal needed stuff. As coqide is built by default I thought to use a more marked naming. We can rename the rule of course..

File: Makefile.dune [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # -*- mode: makefile -*-
 # Dune Makefile for Coq
 
-.PHONY: help voboot states world watch check quickbyte quickopt test-suite release apidoc ocheck ireport clean
+.PHONY: help voboot states world watch check quickbyte quickopt quickide test-suite release apidoc ocheck ireport clean
```
Comment (Zimmi48) [inline Makefile.dune]: I think you should start splitting this list like in `Makefile.dune`.

**Review by Zimmi48 — APPROVED** [review]

**Comment by ejgallego** [comment]
> Building from an empty state, the CoqIDE produced by this target cannot be run as is. Is this expected?

Oh indeed, there are two problems here:
- one is the same problem that `make states` has, [do we have an issue for it] and you must pass `-coqlib $(pwd)/_build/default` to coqide for now. It will be fixed when Dune 1.6 is released, other fixes would be possible as discussed on Gitter, namely adding a smarter logic to `lib/envars`.
- the other one is that the STM doesn't gracefully degrade if the workers are not available, indeed that is a bug on this PR, will fix.

**Comment by ejgallego** [comment]
> * the other one is that the STM doesn't gracefully degrade if the workers are not available, indeed that is a bug on this PR, will fix.

Fixing this properly may require Dune 1.6 too, as we would need to depend on the installed version of the workers; will update the PR nonetheless but this will mean that you can only call coqide manually and not by using `dune exec`.

**Comment by ejgallego** [comment]
New PR pushed, unfortunately it won't work 100% fine until either:
- Dune 1.6 is released,
- or we add smarter rules [not worth doing IMHO].

File: Makefile.dune (hunk 1 above)
Comment (ejgallego) [reply Makefile.dune]: I will do the split in this PR indeed, one sec.

Comment (ejgallego) [reply Makefile.dune]: Well, on the other hand I like the global PHONY as it helps me keeps the help list in sync. I dunno, let me think about it. Sparse phony targets is a problem the current makefile system suffers a lot.

**Comment by Zimmi48** [comment]
When is Dune 1.6 due? If this is soon, then it might make sense to wait instead of merging this now.

**Comment by ejgallego** [comment]
> When is Dune 1.6 due? If this is soon, then it might make sense to wait instead of merging this now.

I'd say before 2019; so do as you prefer indeed, no problem on waiting.

Note these kind of targets [such as `states`] are still useful as you can usually do `world + quickide` and things will work fine. But indeed, that's an ugly hack.

I'd love to hear suggestions about how to handle the location of build-tree artifacts; so far I have been busy with other stuff as to propose a proper design, using the "installed" version seems sensible, but we should do better I think.

**Comment by Zimmi48** [comment]
@gares Do you think we should merge the current patch with its limitations now or wait?

**Comment by gares** [comment]
I think it would be better to wait.

BTW, the `.PHONY:` directive can be repeated, so you can still split the like and have all the PHONY targets be listed on a bunch on adjacent lines. Eg:
```make
.PHONY: a b c d
.PHONY: e f g h
```

**Comment by ejgallego** [comment]
> BTW, the `.PHONY:` directive can be repeated, so you can still split the line and have all the PHONY targets be listed on a bunch on adjacent lines. Eg:

Indeed that is fine to me, I didn't predict the line would grow so long.

**Comment by ejgallego** [comment]
IMO this is ready; the `states` target must be fixed but that'd happen in a separate PR.

**Review by Zimmi48 — CHANGES_REQUESTED** [review]

File: topbin/dune [hunk 2]
```diff
@@ -28,3 +28,8 @@
  (modules :standard \ coqtop_byte_bin coqtop_bin)
  (libraries coq.toplevel)
  (link_flags -linkall))
+
+; Should be changed to depend on the installed targets when Dune 1.6 is released
```
[the remark this answers is not in the document]
Comment (ejgallego) [reply topbin/dune]: Oh indeed, thanks for pointing this out.

[the remark this answers is not in the document]
Comment (Zimmi48) [reply topbin/dune]: ping @ejgallego 
Either fix this or add "needs: progress" / "needs: fixing"

**Comment by ejgallego** [comment]
TTBOMK this is OK now, we still need a shim to get quicker running of the states target, à la #9047 but indeed if this is merged I could update #9047 to add a `coqide-prelude` shim.

**Review by Zimmi48 — APPROVED** [review]

**Comment by Zimmi48** [comment]
Indeed, it would be good to have in addition a trick to be able to run `dune exec coqide` after `make -f Makefile.dune quickide` without having to also run `make -f Makefile.dune world` and kill it after a few seconds. Merging this one now anyway.

**Comment by ejgallego** [comment]
Thanks!

**Pull request merged**